akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/TestKitStorage.scala

```diff
@@ -72,58 +97,74 @@ sealed trait InMemStorage[K, R] extends InternalReprSupport[R] {
   /**
    * Sets new elements returned by updater ordered by seqnum. Sets new seqnum as max(old, max(newElemsFromUpdaterSeqNums))
    */
-  def updateOrSetNew(key: K, updater: Vector[R] => Vector[R]): Vector[R] =
-    eventsMap
-      .compute(
-        key,
-        (_: K, value: (Long, Vector[InternalRepr])) => {
-          val (sn, elems) = if (value != null) value else (0L, Vector.empty)
-          val upd = updater(elems.map(toRepr)).sortBy(reprToSeqNum)
-          (max(getLastSeqNumber(upd), sn), upd.map(toInternal))
-        })
-      ._2
-      .map(toRepr)
+  def updateOrSetNew(key: K, updater: Vector[R] => Vector[R]): Vector[R] = lock.synchronized {
 
-  def read(key: K): Option[Vector[R]] =
-    Option(eventsMap.get(key)).map(_._2.map(toRepr))
+    def remappingFunction: (K, (Long, Vector[InternalRepr])) => (Long, Vector[InternalRepr]) =
+      (_: K, value: (Long, Vector[InternalRepr])) => {
+        val (sn, elems) = if (value != null) value else (0L, Vector.empty)
+        val upd = updater(elems.map(toRepr)).sortBy(reprToSeqNum)
+        (max(getLastSeqNumber(upd), sn), upd.map(toInternal))
+      }
 
-  def readAll(): Iterable[R] = {
-    eventsMap.values().asScala.flatMap { case (_, events) => events }.map(toRepr)
+    val oldValue = eventsMap.getOrElse(key, (0L, Vector.empty))
+    (Option(remappingFunction(key, oldValue)) match {
```

Comment:
Won't this always match `Some` because `remappingFunction` always returns a tuple?

Reply:
👍